### ISSUE TITLE
VIP: allow variable ranges with max bounds

### ISSUE THREAD

**Issue by charles-cooper** [vyperlang/vyper#2527]
## Simple Summary
Add variable ranges with max bounds, e.g. `for i in range(<var>, bound=100)`

## Motivation
Right now vyper requires for-loop ranges to be bounded at compile time. However, an extremely common use case is to iterate over some variable counter at runtime. This leads to the common idiom,
```python
for i in range(SOME_BOUND):
    if i >= self.count:
        break
    <loop body>
```

This proposal allows a runtime variable counter but also requires a max bound on the range.

## Specification
Add a `bound` keyword arg to the `range` builtin, and relax the restrictions on the first argument to allow it to be a non-literal variable. For instance, the above example would become
```python
for i in range(self.count, bound=SOME_BOUND):
    <loop body>
```

## Backwards Compatibility
None known

## Copyright
Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/)

**Comment by charles-cooper** [comment]
This probably becomes more relevant with dynamic arrays https://github.com/vyperlang/vyper/issues/1440

**Comment by fubuloubu** [comment]
Might be more interesting to do it like a slice of the dynamic variable itself:

```py
for item in slice(self.dynamic_array, self.start, LENGTH):
    <loop body>

self.start +=LENGTH
```

**Comment by charles-cooper** [comment]
Yeah often we want access to the `i` value tho, so maybe we need `enumerate(slice(...))`

**Comment by fubuloubu** [comment]
> Yeah often we want access to the `i` value tho, so maybe we need `enumerate(slice(...))`

I like `enumerate`

**Comment by fubuloubu** [comment]
Meeting notes: consider iteration over `slice`, and how it fits in with complex data types #1954

**Comment by charles-cooper** [comment]
and also how it fits in with dynamic arrays https://github.com/vyperlang/vyper/issues/1440

**Comment by charles-cooper** [comment]
Implemented in #3537
